File: python/typelayer_common/typelayer_common/transports/http.py
from .. import application as typelayer_common_application
from .. import dynamic as typelayer_common_dynamic
from ..encodings import json as typelayer_common_encodings_json
from ..encodings import msgpack as typelayer_common_encodings_msgpack
from .. import types as typelayer_common_types
from dataclasses import dataclass
from typing import List, Dict, Optional


class InvalidContentType(typelayer_common_types.TypelayerException):

    def __init__(self, m: str, ):
        super().__init__(m, "typelayer_common_transports_http", "invalid_content_type", {})


@dataclass
class HttpSettings:
    host: str
    port: int

@dataclass
class HttpRequest:
    body: bytes
    headers: Dict[str, str]
    path: List[str]

@dataclass
class HttpResponse:
    body: List[bytes]
    headers: Dict[str, str]
    status_code: int

@dataclass
class HttpState:
    request: "HttpRequest"
    response_headers: Dict[str, str]

class HasHttp:
    def has_http_get(self, ) -> "HttpState":
        raise NotImplementedError
    def has_http_put(self, request: "HttpState") -> None:
        raise NotImplementedError

class HttpApplication:
    def http_application_handle_request(self, request: "HttpRequest") -> "HttpResponse":
        raise NotImplementedError

@dataclass
class BasicHttpApplication(HttpApplication):
    app: typelayer_common_application.Process;
    def http_application_handle_request(self: "BasicHttpApplication", request: "HttpRequest") -> "HttpResponse":
        
        
        encoding = get_encoding_for_request(request)
        result_bin = self.app.process_handle(encoding, request.path, request.body)
        return HttpResponse(status_code=200, headers={"Content-Type": encoding.encoding_mimetype()}, body=[result_bin])

def get_request(has_http_obj: HasHttp) -> "HttpRequest":

    return has_http_obj.has_http_get().request

def get_request_headers(has_http_obj: HasHttp) -> Dict[str, str]:

    return has_http_obj.has_http_get().request.headers

def get_encoding_for_request(request: "HttpRequest") -> typelayer_common_dynamic.Encoding:

    content_type = request.headers.get("content-type", "application/json")
    if (content_type == "application/json"):
        return typelayer_common_encodings_json.JsonEncoding()
    elif (content_type == "application/msgpack"):
        return typelayer_common_encodings_msgpack.MsgPackEncoding()
    else:
            raise InvalidContentType("Invalid Content Type", {});

def put_header(has_http_obj: HasHttp, key: str, value: str) -> None:

    has_http_obj.has_http_get().response_headers[key] = value
    
    has_http_obj.has_http_get().response_headers = has_http_obj.has_http_get().response_headers;
    
    return has_http_obj.has_http_put(has_http_obj.has_http_get())

File: typescript/typelayer-common/transports/http.ts
// Module: transports.http
import * as typelayerCommonApplication from "../application";
import * as typelayerCommonDynamic from "../dynamic";
import * as typelayerCommonEncodingsJson from "../encodings/json";
import * as typelayerCommonEncodingsMsgpack from "../encodings/msgpack";
import * as typelayerCommonTypes from "../types";


export class InvalidContentType extends  typelayerCommonTypes.TypelayerException {

    constructor(m: string, meta: {}) {
        super(m, "typelayer_common_transports_http", "invalid_content_type", new Map([]));
        Object.setPrototypeOf(this, InvalidContentType.prototype);

    }
}

export class HttpSettings {
    host: string;
    port: number;
    constructor(arg: {host: string; port: number;}) {
        Object.assign(this, arg);
    }
}

export class HttpRequest {
    body: Uint8Array;
    headers: Map<string, string>;
    path: Array<string>;
    constructor(arg: {body: Uint8Array; headers: Map<string, string>; path: Array<string>;}) {
        Object.assign(this, arg);
    }
}

export class HttpResponse {
    body: Array<Uint8Array>;
    headers: Map<string, string>;
    statusCode: number;
    constructor(arg: {body: Array<Uint8Array>; headers: Map<string, string>; statusCode: number;}) {
        Object.assign(this, arg);
    }
}

export class HttpState {
    request: HttpRequest;
    responseHeaders: Map<string, string>;
    constructor(arg: {request: HttpRequest; responseHeaders: Map<string, string>;}) {
        Object.assign(this, arg);
    }
}

export interface HasHttp {
    hasHttpGet: () => HttpState;
    hasHttpPut: (request: HttpState) => void;
}

export interface HttpApplication {
    httpApplicationHandleRequest: (request: HttpRequest) => Promise<HttpResponse>;
}

export class BasicHttpApplication implements HttpApplication {
    kind : string = "BasicHttpApplication";
    app: typelayerCommonApplication.Process;
    constructor(arg: {app: typelayerCommonApplication.Process;}) {
        Object.assign(this, arg);
    }
    async httpApplicationHandleRequest(this: BasicHttpApplication, request: HttpRequest): Promise<HttpResponse> {
        let encoding : typelayerCommonDynamic.Encoding;
        let resultBin : Uint8Array;
        encoding = getEncodingForRequest(request);
        resultBin = await this.app.processHandle(encoding, request.path, request.body);
        return new HttpResponse({statusCode: 200, headers: new Map([["Content-Type", encoding.encodingMimetype()]]), body: [resultBin]});
    }
}

export function getRequest(hasHttpObj: HasHttp): HttpRequest {

    return hasHttpObj.hasHttpGet().request;
}

export function getRequestHeaders(hasHttpObj: HasHttp): Map<string, string> {

    return hasHttpObj.hasHttpGet().request.headers;
}

export function getEncodingForRequest(request: HttpRequest): typelayerCommonDynamic.Encoding {
    let contentType : string;
    contentType = (request.headers.get("content-type") || "application/json");
    switch (contentType) {
        case "application/json":
            return new typelayerCommonEncodingsJson.JsonEncoding({});
            break;
        case "application/msgpack":
            return new typelayerCommonEncodingsMsgpack.MsgPackEncoding({});
            break;
        default:
            throw new InvalidContentType("Invalid Content Type", {});
    }
}

export function putHeader(hasHttpObj: HasHttp, key: string, value: string): void {

    hasHttpObj.hasHttpGet().responseHeaders.set(key, value);
    
    hasHttpObj.hasHttpGet().responseHeaders = hasHttpObj.hasHttpGet().responseHeaders;
    
    return hasHttpObj.hasHttpPut(hasHttpObj.hasHttpGet());
}